proposal/nested-map-functions.md

```diff
@@ -0,0 +1,204 @@
+
+## Table of Contents
+
+* [Background](#background)
+* [Summary](#summary)
+* [Syntax](#syntax)
+  * [`get()`](#get)
+  * [`has-key()`](#has-key)
+  * [`set()`](#set)
+  * [`merge()`](#merge)
+
+## Background
+
+> This section is non-normative.
+
+The current map inspection and manipulation functions don't provide any built-in
+support for managing nested maps. Projects often build thir own tooling, but
+the results are inconsistent, and often slow.
+
+## Summary
+
+> This section is non-normative.
+
+This proposal updates the `sass:map` module to better support inspection and
+manipulation of nested maps.
+
+The `has-key()` and `get()` functions both accept multiple `$keys...`:
+
+```scss
+@use 'sass:map';
+
+$nav: (
+  'bg': 'gray',
+  'color': (
+    'hover': (
+      'search': yellow,
+      'home': red,
+      'filter': blue,
+    ),
+  ),
+);
+
+$has-search: map.has-key($nav, 'color', 'hover', 'search'); // true
+$search-hover: map.get($nav, 'color', 'hover', 'search'); // yellow
+```
+
+The `merge()` function now accepts multiple `$keys...` between the two maps
+being merged. The keys form a path to the nested location in `$map1` where
+`$map2` should be merged. For example, we update the hover colors in our `$nav`
+map above:
+
+```scss
+@use 'sass:map';
+
+$new-hover: (
+  'search': green,
+  'logo': orange,
+);
+
+$nav: map.merge($nav, 'color', 'hover', $new-hover);
+
+// $nav: (
+//   'bg': 'gray',
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+This proposal also adds a `set()` function to `sass:map`, with a similar syntax,
+returning a map with any nested key set to a specific value. To achieve the
+same output as our merge example, we can set each key individually:
+
+```scss
+@use 'sass:map';
+
+$nav: map.set($nav, 'color', 'hover', 'search', green);
+$nav: map.set($nav, 'color', 'hover', 'logo', orange);
+```
+
+## Syntax
+
+### `get()`
+
+```
+get($map, $keys...)
+```
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key with the same name as `key`, throw an error.
+
+* Let `value` be the value assigned to `key` in `$map`
+
+* If there is more than one element in `$keys`:
+
+  * Let `keys` be all elements in `$keys` after the first element
+
+  * Call `get()` with `value` and expanded `keys` as arguments
+
+* Otherwise, return `value`
+
+### `has-key()`
+
+```
+has-key($map, $keys...)
+```
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key with the same name as `key`, return boolean
+  `false`.
+
+* If there is more than one element in `$keys`:
+
+  * Let `map` be the value assigned to `key` in `$map`
+
+  * Let `keys` be all but the first element in `$keys`
+
+  * Call `has-key()` with `map` and expanded `keys` as arguments
+
+* Otherwise, return boolean `true`
+
+### `set()`
+
+```
+set($map1, $keys..., $value)
```

Comment:
Sass doesn't natively allow var args to be written before a non-variable argument, so this may require a little more machinery than usual to make that work. I don't think we need to make `map.set($map: ..., $value: ...)` work for a new function, but you may need to write the signature as `map.set($map, $args...)` and explicitly slice out the value from the argument list.